docs/configuration.md
Configuration
=============

Pitaya uses Viper to control its configuration. Below we describe the configuration variables split by topic. We judge the default values are good for most cases, but might need to be changed for some use cases.

## Service Discovery

These configuration values configure service discovery for the default etcd service discovery module.
 They only need to be set if the application runs in cluster mode.

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.cluster.sd.etcd.dialtimeout | 5s | time.Time | Dial timeout value passed to the service discovery etcd client |
| pitaya.cluster.sd.etcd.endpoints | localhost:2379 | string | List of comma separated etcd endpoints |
| pitaya.cluster.sd.etcd.heartbeat.ttl | 60s | time.Time | Hearbeat interval for the etcd lease |
| pitaya.cluster.sd.etcd.heartbeat.log | false | bool | Whether etcd heartbeats should be logged in debug mode |
| pitaya.cluster.sd.etcd.prefix | pitaya/ | string | Prefix used to avoid collisions with different pitaya applications, servers must have the same prefix to be able to see each other |
| pitaya.cluster.sd.etcd.syncservers.interval | 120s | time.Time | Interval between server syncs performed by the service discovery module |

## RPC Service

The configurations only need to be set if the RPC Service is enabled with the given type.

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.buffer.cluster.rpc.server.nats.messages | 75 | int | Size of the buffer that for the nats RPC server accepts before starting to drop incoming messages |
| pitaya.buffer.cluster.rpc.server.nats.push | 100 | int | Size of the buffer that the nats RPC server creates for push messages |
| pitaya.cluster.rpc.client.grpc.requesttimeout | 5s | time.Time | Request timeout for RPC calls with the gRPC client |
| pitaya.cluster.rpc.client.grpc.dialtimeout | 5s | time.Time| Timeout for the gRPC client to establish the connection |
| pitaya.cluster.rpc.client.nats.connect | nats://localhost:4222 | string | Nats address for the client |
| pitaya.cluster.rpc.client.nats.requesttimeout | 5s | time.Time | Request timeout for RPC calls with the nats client |
| pitaya.cluster.rpc.client.nats.maxreconnectionretries | 15 | int | Maximum number of retries to reconnect to nats for the client |
| pitaya.cluster.rpc.server.nats.connect | nats://localhost:4222 | string | Nats address for the server |
| pitaya.cluster.rpc.server.nats.maxreconnectionretries | 15 | int | Maximum number of retries to reconnect to nats for the server |
| pitaya.cluster.rpc.server.grpc.port | 3434 | int | The port that the gRPC server listens to |
| pitaya.concurrency.remote.service | 30 | int | Number of goroutines processing messages at the remote service for the nats RPC service |

## Connection

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.handler.messages.compression | true | bool | Whether messages between client and server should be compressed |
| pitaya.heartbeat.interval | 30s | time.Time | Keepalive heartbeat interval for the client connection |

## Metrics Reporting

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.metrics.statsd.enabled | false | bool | Whether statsd reporting should be enabled |
| pitaya.metrics.statsd.host | localhost:9125 | string | Address of the statsd server to send the metrics to |
| pitaya.metrics.statsd.prefix | pitaya. | string | Prefix of the metrics reported to statsd |
| pitaya.metrics.statsd.rate | 1 | int | Statsd metrics rate |
| pitaya.metrics.prometheus.enabled | false | bool | Whether prometheus reporting should be enabled |
| pitaya.metrics.prometheus.port | 9090 | int | Port to expose prometheus metrics |
| pitaya.metrics.tags | map[string]string{} | map[string]string | Tags to be added to reported metrics |

## Concurrency

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.buffer.agent.messages | 100 | int | Buffer size for received client messages for each agent |
| pitaya.buffer.handler.localprocess | 20 | int | Buffer size for messages received by the handler and processed locally |
| pitaya.buffer.handler.remoteprocess | 20 | int | Buffer size for messages received by the handler and forwarded to remote servers |
| pitaya.concurrency.handler.dispatch | 25 | int | Number of goroutines processing messages at the handler service |

## Modules

These configurations are only used if the modules are created. It is recommended to use Binding Storage module with gRPC RPC service to be able to use all RPC service features.

| Configuration | Default value | Type | Description |
|------------------------------------|---------------|-----------|------------------------------------------|
| pitaya.session.unique | true | bool | Whether Pitaya should enforce unique sessions for the clients, enabling the unique sessions module |
| pitaya.modules.bindingstorage.etcd.endpoints | localhost:2379 | string | Comma separated list of etcd endpoints to be used by the binding storage module, should be the same as the service discovery etcd |
| pitaya.modules.bindingstorage.etcd.prefix | pitaya/ | string | Prefix used for etcd, should be the same as the service discovery |
| pitaya.modules.bindingstorage.etcd.dialtimeout | 5s | time.Time | Timeout to establish the etcd connection |
| pitaya.modules.bindingstorage.etcd.leasettl | 1h | time.Time | Duration of the etcd lease before automatic renewal |



docs/configuration.rst
*************
Configuration
*************

Pitaya uses Viper to control its configuration. Below we describe the configuration variables split by topic. We judge the default values are good for most cases, but might need to be changed for some use cases.

Service Discovery
=================

These configuration values configure service discovery for the default etcd service discovery module.
 They only need to be set if the application runs in cluster mode.

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.cluster.sd.etcd.dialtimeout
    - 5s
    - time.Time
    - Dial timeout value passed to the service discovery etcd client
  * - pitaya.cluster.sd.etcd.endpoints
    - localhost:2379
    - string
    - List of comma separated etcd endpoints
  * - pitaya.cluster.sd.etcd.heartbeat.ttl
    - 60s
    - time.Time
    - Hearbeat interval for the etcd lease
  * - pitaya.cluster.sd.etcd.heartbeat.log
    - false
    - bool
    - Whether etcd heartbeats should be logged in debug mode
  * - pitaya.cluster.sd.etcd.prefix
    - pitaya/
    - string
    - Prefix used to avoid collisions with different pitaya applications, servers must have the same prefix to be able to see each other
  * - pitaya.cluster.sd.etcd.syncservers.interval
    - 120s
    - time.Time
    - Interval between server syncs performed by the service discovery module

RPC Service
===========

The configurations only need to be set if the RPC Service is enabled with the given type.

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.buffer.cluster.rpc.server.nats.messages
    - 75
    - int
    - Size of the buffer that for the nats RPC server accepts before starting to drop incoming messages
  * - pitaya.buffer.cluster.rpc.server.nats.push
    - 100
    - int
    - Size of the buffer that the nats RPC server creates for push messages
  * - pitaya.cluster.rpc.client.grpc.requesttimeout
    - 5s
    - time.Time
    - Request timeout for RPC calls with the gRPC client
  * - pitaya.cluster.rpc.client.grpc.dialtimeout
    - 5s
    - time.Tim
    - Timeout for the gRPC client to establish the connection
  * - pitaya.cluster.rpc.client.nats.connect
    - nats://localhost:4222
    - string
    - Nats address for the client
  * - pitaya.cluster.rpc.client.nats.requesttimeout
    - 5s
    - time.Time
    - Request timeout for RPC calls with the nats client
  * - pitaya.cluster.rpc.client.nats.maxreconnectionretries
    - 15
    - int
    - Maximum number of retries to reconnect to nats for the client
  * - pitaya.cluster.rpc.server.nats.connect
    - nats://localhost:4222
    - string
    - Nats address for the server
  * - pitaya.cluster.rpc.server.nats.maxreconnectionretries
    - 15
    - int
    - Maximum number of retries to reconnect to nats for the server
  * - pitaya.cluster.rpc.server.grpc.port
    - 3434
    - int
    - The port that the gRPC server listens to
  * - pitaya.concurrency.remote.service
    - 30
    - int
    - Number of goroutines processing messages at the remote service for the nats RPC service

Connection
==========

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.handler.messages.compression
    - true
    - bool
    - Whether messages between client and server should be compressed
  * - pitaya.heartbeat.interval
    - 30s
    - time.Time
    - Keepalive heartbeat interval for the client connection

Metrics Reporting
=================

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.metrics.statsd.enabled
    - false
    - bool
    - Whether statsd reporting should be enabled
  * - pitaya.metrics.statsd.host
    - localhost:9125
    - string
    - Address of the statsd server to send the metrics to
  * - pitaya.metrics.statsd.prefix
    - pitaya.
    - string
    - Prefix of the metrics reported to statsd
  * - pitaya.metrics.statsd.rate
    - 1
    - int
    - Statsd metrics rate
  * - pitaya.metrics.prometheus.enabled
    - false
    - bool
    - Whether prometheus reporting should be enabled
  * - pitaya.metrics.prometheus.port
    - 9090
    - int
    - Port to expose prometheus metrics
  * - pitaya.metrics.tags
    - map[string]string{}
    - map[string]string
    - Tags to be added to reported metrics

Concurrency
===========

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.buffer.agent.messages
    - 100
    - int
    - Buffer size for received client messages for each agent
  * - pitaya.buffer.handler.localprocess
    - 20
    - int
    - Buffer size for messages received by the handler and processed locally
  * - pitaya.buffer.handler.remoteprocess
    - 20
    - int
    - Buffer size for messages received by the handler and forwarded to remote servers
  * - pitaya.concurrency.handler.dispatch
    - 25
    - int
    - Number of goroutines processing messages at the handler service

Modules
=======

These configurations are only used if the modules are created. It is recommended to use Binding Storage module with gRPC RPC service to be able to use all RPC service features.

.. list-table::
  :widths: 15 10 10 50
  :header-rows: 1
  :stub-columns: 1

  * - Configuration
    - Default value
    - Type
    - Description
  * - pitaya.session.unique
    - true
    - bool
    - Whether Pitaya should enforce unique sessions for the clients, enabling the unique sessions module
  * - pitaya.modules.bindingstorage.etcd.endpoints
    - localhost:2379
    - string
    - Comma separated list of etcd endpoints to be used by the binding storage module, should be the same as the service discovery etcd
  * - pitaya.modules.bindingstorage.etcd.prefix
    - pitaya/
    - string
    - Prefix used for etcd, should be the same as the service discovery
  * - pitaya.modules.bindingstorage.etcd.dialtimeout
    - 5s
    - time.Time
    - Timeout to establish the etcd connection
  * - pitaya.modules.bindingstorage.etcd.leasettl
    - 1h
    - time.Time
    - Duration of the etcd lease before automatic renewal



config/config.go
// Copyright (c) TFG Co. All Rights Reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package config

import (
	"strings"
	"time"

	"github.com/spf13/viper"
)

// Config is a wrapper around a viper config
type Config struct {
	config *viper.Viper
}

// NewConfig creates a new config with a given viper config if given
func NewConfig(cfgs ...*viper.Viper) *Config {
	var cfg *viper.Viper
	if len(cfgs) > 0 {
		cfg = cfgs[0]
	} else {
		cfg = viper.New()
	}

	cfg.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	cfg.AutomaticEnv()
	c := &Config{config: cfg}
	c.fillDefaultValues()
	return c
}

func (c *Config) fillDefaultValues() {
	defaultsMap := map[string]interface{}{
		"pitaya.buffer.agent.messages": 100,
		// the max buffer size that nats will accept, if this buffer overflows, messages will begin to be dropped
		"pitaya.buffer.cluster.rpc.server.nats.messages": 75,
		"pitaya.buffer.cluster.rpc.server.nats.push":     100,
		"pitaya.buffer.handler.localprocess":             20,
		"pitaya.buffer.handler.remoteprocess":            20,
		// the sum of this config among all the frontend servers should always be less than
		// the sum of pitaya.buffer.cluster.rpc.server.nats.messages, for covering the worst case scenario
		// a single backend server should have the config pitaya.buffer.cluster.rpc.server.nats.messages bigger
		// than the sum of the config pitaya.concurrency.handler.dispatch among all frontend servers
		"pitaya.concurrency.handler.dispatch":                   25,
		"pitaya.concurrency.remote.service":                     30,
		"pitaya.cluster.rpc.client.grpc.requesttimeout":         "5s",
		"pitaya.cluster.rpc.client.grpc.dialtimeout":            "5s",
		"pitaya.cluster.rpc.client.nats.connect":                "nats://localhost:4222",
		"pitaya.cluster.rpc.client.nats.requesttimeout":         "5s",
		"pitaya.cluster.rpc.client.nats.maxreconnectionretries": 15,
		"pitaya.cluster.rpc.server.nats.connect":                "nats://localhost:4222",
		"pitaya.cluster.rpc.server.nats.maxreconnectionretries": 15,
		"pitaya.cluster.rpc.server.grpc.port":                   3434,
		"pitaya.cluster.sd.etcd.dialtimeout":                    "5s",
		"pitaya.cluster.sd.etcd.endpoints":                      "localhost:2379",
		"pitaya.cluster.sd.etcd.prefix":                         "pitaya/",
		"pitaya.cluster.sd.etcd.heartbeat.ttl":                  "60s",
		"pitaya.cluster.sd.etcd.heartbeat.log":                  false,
		"pitaya.cluster.sd.etcd.syncservers.interval":           "120s",
		"pitaya.modules.bindingstorage.etcd.endpoints":          "localhost:2379",
		"pitaya.modules.bindingstorage.etcd.prefix":             "pitaya/",
		"pitaya.modules.bindingstorage.etcd.dialtimeout":        "5s",
		"pitaya.modules.bindingstorage.etcd.leasettl":           "1h",
		"pitaya.handler.messages.compression":                   true,
		"pitaya.heartbeat.interval":                             "30s",
		"pitaya.session.unique":                                 true,
		"pitaya.metrics.statsd.enabled":                         false,
		"pitaya.metrics.statsd.host":                            "localhost:9125",
		"pitaya.metrics.statsd.prefix":                          "pitaya.",
		"pitaya.metrics.statsd.rate":                            1,
		"pitaya.metrics.prometheus.port":                        9090,
		"pitaya.metrics.prometheus.enabled":                     false,
		"pitaya.metrics.tags":                                   map[string]string{},
	}

	for param := range defaultsMap {
		if c.config.Get(param) == nil {
			c.config.SetDefault(param, defaultsMap[param])
		}
	}
}

// GetDuration returns a duration from the inner config
func (c *Config) GetDuration(s string) time.Duration {
	return c.config.GetDuration(s)
}

// GetString returns a string from the inner config
func (c *Config) GetString(s string) string {
	return c.config.GetString(s)
}

// GetInt returns an int from the inner config
func (c *Config) GetInt(s string) int {
	return c.config.GetInt(s)
}

// GetBool returns an boolean from the inner config
func (c *Config) GetBool(s string) bool {
	return c.config.GetBool(s)
}

// GetStringSlice returns a string slice from the inner config
func (c *Config) GetStringSlice(s string) []string {
	return c.config.GetStringSlice(s)
}

// Get returns an interface from the inner config
func (c *Config) Get(s string) interface{} {
	return c.config.Get(s)
}

// GetStringMapString returns a string map string from the inner config
func (c *Config) GetStringMapString(s string) map[string]string {
	return c.config.GetStringMapString(s)
}
